rpc_util.go

```diff
@@ -148,7 +148,9 @@ func encode(msg proto.Message, pf payloadFormat) ([]byte, error) {
 		}
 		length = uint32(len(b))
 	}
-	if err := binary.Write(&buf, binary.BigEndian, length); err != nil {
+	var szHdr [4]byte
+	binary.BigEndian.PutUint32(szHdr[:], length)
+	if _, err := buf.Write(szHdr[:]); err != nil {
```

Comment:
Sure!  In the original, you are using `binary.Write`([binary.Write impl.](http://golang.org/src/encoding/binary/binary.go?s=6251:8382#L220)).  Each time a message is encoded, the `binary.Write` function body needs to …

1. allocate two byte arrays and one slice (three allocations in on heap in total)
2. perform [type switching](https://golang.org/doc/effective_go.html#type_switch)

Admittedly the compiler and runtime could improve the memory behavior (keep on stack) and the type switch through some sort of a priori branch inferencing in future versions, but the `encoding/binary` package [explicitly states](http://golang.org/src/encoding/binary/binary.go#L18):

> This package favors simplicity over efficiency.

… whereas we know a priori the storage size of the length (four bytes), which means …

1. we can specify the [VarDecl](https://golang.org/ref/spec#VarDecl) with that size once and hold the number of allocations to one
2. we specify the exact encoding mechanism `binary.BigEndian.PutUint32` as opposed to being forced by the runtime to type switch

… without those overheads.  The net effect of this is faster encoding, reduced memory usage (size), and reduced heap allocation (object count), all of which mean increased throughput (bytes/second).  With this being library code, the memory stewardship component is salient.

In the commit message, I excerpt the result of this (before versus after) using the [benchcmp tool](https://github.com/golang/tools/blob/master/cmd/benchcmp/benchcmp.go):

```
$ # Return to previous commit but benchmark.
$ go test ./... -test.bench="Benchmark*" > /tmp/before
$ # Return to working copy.
$ go test ./... -test.bench="Benchmark*" > /tmp/after
$ benchcmp /tmp/before /tmp/after
benchmark               old ns/op     new ns/op     delta
BenchmarkEncode1        1451          942           -35.08%
BenchmarkEncode10       1466          1067          -27.22%
BenchmarkEncode100      1966          1648          -16.17%
BenchmarkEncode1000     4390          3910          -10.93%

benchmark               old MB/s     new MB/s     speedup
BenchmarkEncode1        5.51         8.49         1.54x
BenchmarkEncode10       11.60        15.93        1.37x
BenchmarkEncode100      54.40        64.92        1.19x
BenchmarkEncode1000     229.59       257.77       1.12x

benchmark               old allocs     new allocs     delta
BenchmarkEncode1        6              3              -50.00%
BenchmarkEncode10       7              4              -42.86%
BenchmarkEncode100      8              5              -37.50%
BenchmarkEncode1000     8              5              -37.50%

benchmark               old bytes     new bytes     delta
BenchmarkEncode1        384           328           -14.58%
BenchmarkEncode10       400           344           -14.00%
BenchmarkEncode100      736           680           -7.61%
BenchmarkEncode1000     2560          2504          -2.19%
```